### PR TITLE
Issue #19717: Add input files for FinalLocalVariableCheck about initialization in adjacent catches

### LINKED ISSUE #19717: False Negative: FinalLocalVariableCheck does not report for variables assigned in adjacent catches
I have read check documentation: https://checkstyle.org/checks/xxxxxx/nameofaffectedcheck.html
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```bash
/var/tmp $ javac Test.java 
/var/tmp $ cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
    <module name="TreeWalker">
        <module name="FinalLocalVariableCheck"/>
    </module>
</module>
/var/tmp $ cat Test.java 
import java.io.FileNotFoundException;
import java.nio.channels.FileLock;

public class Test {
    public void test(boolean x) {
        int u;  // violation, but not reported
        try {
        } catch (IndexOutOfBoundsException e) {
            u = 9;
        } catch (RuntimeException e) {
            u = 10;
        }

        int v;  // violation
        if (x) {
            v = 0;
        } else {
            v = 1;
        }
    }

}
/var/tmp $ RUN_LOCALE="-Duser.language=en -Duser.country=US"
/var/tmp $ java $RUN_LOCALE -jar checkstyle-13.4.0-all.jar -c config.xml Test.java 
Starting audit...
[ERROR] /mnt/d/IntellijProjects/checkstyle-playground/inputs/final-local-variable/Test.java:14:13: Variable 'v' should be declared final. [FinalLocalVariable]
Audit done.
Checkstyle ends with 1 errors.
```
---

**Describe what you expect in detail.**

A violation on `u` not declared final should be reported, since catch clauses are mutually-exclusive.

### PR DESCRIPTION
**Pull request by lithops-zty**
fixes: #19717

Fix false negative in `FinalLocalVariableCheck` for multi-catch blocks.

Problem
---

When multiple `catch` blocks each assign the same uninitialized variable, the variable should be reported as a candidate for `final` because only one `catch` branch can execute, so the variable is effectively assigned exactly once on every path. Currently, the check silently misses this case.

Root Cause
---

Two methods in FinalLocalVariableCheck lack awareness of LITERAL_CATCH as a mutually-exclusive branch (analogous to LITERAL_ELSE for if-else and SWITCH_RULE for arrow-switch):

1. determineAssignmentConditions: its blockTypes array does not include LITERAL_CATCH, so when the second catch block's assignment is processed and candidate.assigned is already true, the check unconditionally sets candidate.alreadyAssigned = true and removes the variable from the final-candidate list.
2. shouldUpdateUninitializedVariables: when leaving the first catch block's SLIST, the method does not recognize that another catch follows, so updateAllUninitializedVariables() is never called and the variable's uninitialized state is lost before the second catch is visited.

Fix
---
1. Add TokenTypes.LITERAL_CATCH to the blockTypes array in determineAssignmentConditions.
2. Add a new helper isCatchTokenWithAnotherCatchFollowing (symmetric with the existing isCaseTokenWithAnotherCaseFollowing) and call it from shouldUpdateUninitializedVariables:

### CODE REVIEW

**Comment by lithops-zty** [comment]
Github, generate report for FinalLocalVariableCheck/all-examples-in-one

File: config/checkstyle-checks.xml [hunk 1]
```diff
@@ -924,7 +924,7 @@
     </module>
     <module name="LambdaBodyLength"/>
     <module name="MethodCount">
-      <property name="maxTotal" value="34"/>
+      <property name="maxTotal" value="35"/>
```
Comment (lithops-zty) [inline config/checkstyle-checks.xml:927]: The value 34 was chosen exactly because this check had 34 methods and was the largest of all checks: (#4539 (comment))
I think it is reasonable to increase it as the updated version now has 35 methods.

**Comment by lithops-zty** [comment]
@romani PR ready for review.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/finallocalvariable/InputFinalLocalVariableMultiCatch.java [hunk 2]
```diff
@@ -10,12 +10,71 @@
 
 public class InputFinalLocalVariableMultiCatch {
     public void demo() throws Throwable {
+        // violation below, "Variable 'a' should be declared final"
+        int a; // assigned in adjacent catches
```
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/finallocalvariable/InputFinalLocalVariableMultiCatch.java]: > assigned in adjacent catches

assigned in nested catches

------


very hard to read such Inputs, yes, it is very comprehensive and compact, but it will not make it good in long run.

lets make them compact in formatting, but keep single variable to validate and right under correct form: 
```
       int u;  // violation
        try { } catch (IndexOutOfBoundsException e) { u = 9; } 

       final int u;
        try { } catch (IndexOutOfBoundsException e) { u = 9; } 

```

**Review by romani — CHANGES_REQUESTED** [review]
thanks a lot, minor:

File: config/checkstyle-checks.xml (hunk 1 above)
Comment (romani) [reply config/checkstyle-checks.xml]: you can put permanent suppression on this Check + File to our suppression xml, there are already bunch like this.
We do not know how to design classes to be less big and same time easy to read and keep logic very close.

**Comment by romani** [comment]
Github, generate report for FinalLocalVariableCheck/all-examples-in-one

**Comment by romani** [comment]
Github, generate report for FinalLocalVariable/all-examples-in-one

**Comment by romani** [comment]
@lithops-zty , small update please and we will merge

File: config/checkstyle-checks.xml (hunk 1 above)
Comment (lithops-zty) [reply config/checkstyle-checks.xml]: done. I restored to 34 and added suppression

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/finallocalvariable/InputFinalLocalVariableMultiCatch.java (hunk 2 above)
Comment (lithops-zty) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/finallocalvariable/InputFinalLocalVariableMultiCatch.java]: That's true. I've completely separated each case into different try-catch blocks and added the correct version right under each violation.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/finallocalvariable/InputFinalLocalVariableMultiCatchAdjacentCatches.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        } catch (IllegalStateException | NullPointerException e) {
+            aCorrect = 1;
+        } catch (final IllegalArgumentException e) {
+            aCorrect = 2;
+            final int x = aCorrect;
+        }
+
+        byte b;
+        try {
+            b = -1;
```
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/finallocalvariable/InputFinalLocalVariableMultiCatchAdjacentCatches.java:41]: Why no violation ?

**Review by romani — CHANGES_REQUESTED** [review]
Not clear why some violations are missed

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/finallocalvariable/InputFinalLocalVariableMultiCatchAdjacentCatches.java (hunk 3 above)
Comment (lithops-zty) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/finallocalvariable/InputFinalLocalVariableMultiCatchAdjacentCatches.java:41]: In this case `b` is assigned in both `try` and `catch`. 
It is very hard to say whether assignment will happen only once in `try`, as we cannot determine syntactically if exceptions will be thrown.
In fact, java disallow `final` in this case:
```
(base) zty@zty:/mnt/d/IntellijProjects/checkstyle-playground/inputs/final-local-variable$ cat Test.java 
public class Test {
    public void test() {
        final byte b;
        try {
            b = -1;
        } catch (final NumberFormatException e) {
            b = 0;
        } catch (IllegalStateException | NullPointerException e) {
            b = 0;
        }
    }

}(base) zty@zty:/mnt/d/IntellijProjects/checkstyle-playground/inputs/final-local-variable$ javac Test.java 
Test.java:7: error: variable b might already have been assigned
            b = 0;
            ^
Test.java:9: error: variable b might already have been assigned
            b = 0;
            ^
2 errors
```

Comment (romani) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/finallocalvariable/InputFinalLocalVariableMultiCatchAdjacentCatches.java:41]: ok, but why this violation:
```
        int a; // violation, "Variable 'a' should be declared final"
        try {
        } catch (final NumberFormatException e) {
            a = 0;
        } catch (IllegalStateException | NullPointerException e) {
            a = 1;
        } catch (final IllegalArgumentException e) {
            a = 2;
        }
```

as if no exception happens, `a` is not initialized.

if this is defect of compiler, we should not put violation on this, as eventually it will be fixed in compiler.
Can you looks in jdk for defect on this? if not present you can create :)

Comment (lithops-zty) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/finallocalvariable/InputFinalLocalVariableMultiCatchAdjacentCatches.java:41]: I am not sure if this is a defect, but Java allows uninitialized final variable declaration. However, such variables cannot be read later on and will lead to a compiling error:
```
(base) zty@zty:/mnt/d/IntellijProjects/checkstyle-playground/inputs/final-local-variable$ cat Test.java 
public class Test {
    public void test() {
        final int a;
        try {
        } catch (final NumberFormatException e) {
            a = 0;
        } catch (IllegalStateException | NullPointerException e) {
            a = 1;
        } catch (final IllegalArgumentException e) {
            a = 2;
        }
        int b = a;
    }

}(base) zty@zty:/mnt/d/IntellijProjects/checkstyle-playground/inputs/final-local-variable$ javac Test.java  
Test.java:12: error: variable a might not have been initialized
        int b = a;
                ^
1 error
```

Comment (romani) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/finallocalvariable/InputFinalLocalVariableMultiCatchAdjacentCatches.java:41]: Something tells me that we should not try to validate this at all, to much uncertainty, even compiler is not consistent.

**Comment by romani** [comment]
I more and inclined to not do violations in such code.

We would rather do less violations, to not produce false positives. False positives blocks builds and annoy a lot user. Missed violations are ok. Missed violations on crazy edges are ok.

I recommend  to remove functional changes and keep all test extension to make it clear we considered it but decided to not do.

If we in agreement, I will change issue title

**Comment by lithops-zty** [comment]
Sorry for replying late.
I agree this is an edge case that may not warrant strict enforcement.
But do you think there's merit to leave this an option to users, i.e. add a property to trigger it?
That said, I am OK to remove it completely.

**Comment by romani** [comment]
I don't see value.
Users usually lazy to look at docs for property, they simply disable Check if it misbehaving or blocking them.

Let's in this PR remove all changes in check, and simply merge Inputs, and we can add some comment over variables that until compiler becomes strict, we should not place violation. And we can merge PR.

Can you create issue over java compiler? In worse case they close it with explanation of reason, that is also good to know.

**Comment by lithops-zty** [comment]
No problem, I will remove all functional changes.

However, I don't see a problem with the compiler, could you elaborate on the exact bug you mentioned?

To me, the compiler works as I would expect in terms of allowing and disallowing `final`:
1. **assignments in both try and catch**
```java
final byte b;  // compile error
try {
    b = -1;
} catch (final NumberFormatException e) {
    b = 0;
}
```
`final` is forbidden because `try` and `catch` are not mutually exclusive and both assignments may execute.
2. **assignments in catches only**
```java
final int a;
try {
} catch (final NumberFormatException e) {
    a = 0;
} catch (IllegalStateException | NullPointerException e) {
    a = 1;
}
```
`final` is allowed because `catch`es are mutually exclusive and at most one assignment may execute.

**Comment by romani** [comment]
>assignments in both try and catc

Yes, we should not make violation on such code.

>2. assignments in catches only

If exception is not happening, what value will have a variable?
And there might be some method in try that throw RuntimeExceptoon , We can guarantee initialization only by `catch Throwable`. There might be a lot of nuances, but still checkstyle should not violate this. Nobody writes code like this, so it might be purely academic problem :).

**Comment by lithops-zty** [comment]
@romani Thanks for the insights. Now I see there is no value checking case 2 since such variable cannot be used at all despite successful compilation. I have pushed a new commit with only input files.

**Review by romani — APPROVED** [review]
Thanks a lot
